### ISSUE TITLE
Second converge fails when using chef-vault because clients dir is wiped out

### ISSUE THREAD

**Issue by jf647** [test-kitchen/test-kitchen#602]
# TL;DR

Incoming wall of text.  The fix is simple, but as there isn't enough data as to why the original change was made, I want to float this by @jtimberman and @fnichol first.

Back in [this commit](https://github.com/test-kitchen/test-kitchen/commit/27f3a3db2b2c2387808d6aaa9140ffc65e9a538a#diff-1873d51ac47ab998a55c290816b6f98a), changes were made to nuke and re-build the clients dir inside of /tmp/kitchen before converging.

This breaks my ability to run converge a cookbook that uses chef-vault more than once because the dynamically created client only has a public key during the first run:

``` bash
Recipe: chefci_tk_secrets::create_vaults
  * ruby_block[create vaults] action create
    - execute the ruby block create vaults
```

But not during the second:

``` bash
Recipe: chefci_tk_secrets::create_vaults
  * ruby_block[create vaults] action create
           ================================================================================
           Error executing action `create` on resource 'ruby_block[create vaults]'
           ================================================================================

           ChefVault::Exceptions::ClientNotFound
           -------------------------------------
           default-ubuntu-1404 is not a valid chef client and/or node
```
# Background

chef-vault works by symmetrically encrypting a data bag item and then creating a separate asymmetrically encrypted copy of the shared secret using the Chef public key for some number of admin users and clients (typically returned from a Chef SOLR query).

When testing a cookbook that uses chef-vault, you can't pre-encrypt your secrets because the keypair of the node under test is generated on the fly during the first converge.

The way around this is to use a fixture cookbook that only gets added to the node's run list when being run under test kitchen.  The fixture cookbook creates a vault so that the real recipe (which comes after the fixture cookbook in the kitchen runlist) can use the secrets:

``` yaml
suites:
  - name: default
    run_list:
      - recipe[testsetup]
      - recipe[nord_rabbitmq]
```

And then `text/fixtures/cookbooks/testsetup/recipes/default.rb` might contain this:

``` ruby
vi = ChefVault::Item.new('foo', 'bar')
vi['rmq_user'] = 'rabbit'
vi['rmq_pass'] = 'sekrit'
vi.clients('*:*')
vi.save
```

(for the curious, you use Berkshelf to get the fixture cookbook into the testing VM):

``` ruby
group :integration do
  cookbook 'testsetup', path: 'test/fixtures/cookbooks/testsetup'
end
```
# Problem

During the first converge, there is no `client.pem` file, so chef-client registers, creating a client object and a node object.  If you poke around after the first converge, you can inspect these:

``` bash
vagrant@default-ubuntu-1404:/tmp/kitchen$ pwd
/tmp/kitchen
vagrant@default-ubuntu-1404:/tmp/kitchen$ ls
backup  cache  client.pem  client.rb  clients  cookbooks  data_bags  dna.json  environments  nodes
vagrant@default-ubuntu-1404:/tmp/kitchen$ ls -la clients nodes
clients:
total 12
-rw-r--r-- 1 root    root     515 Feb  6 16:14 default-ubuntu-1404.json

nodes:
total 52
-rw-r--r-- 1 root    root    41900 Feb  6 16:17 default-ubuntu-1404.json
```

And indeed, the chef-vault data bags have been created properly by the fixture cookbook:

``` bash
vagrant@default-ubuntu-1404:/tmp/kitchen$ ls -la data_bags/rabbitmq
total 24
-rw-r--r-- 1 root root  918 Feb  6 16:15 prod.json
-rw-r--r-- 1 root root  500 Feb  6 16:15 prod_keys.json
-rw-r--r-- 1 root root  918 Feb  6 16:15 test.json
-rw-r--r-- 1 root root  500 Feb  6 16:15 test_keys.json
```

But let's assume that the first converge fails because of a problem with a recipe after the fixture cookbook has run.  You fix the error and converge again.  Prior to the second converge, test-kitchen [wipes out the clients dir](https://github.com/test-kitchen/test-kitchen/blob/master/lib/kitchen/provisioner/chef_base.rb#L94):

``` bash
-----> Starting Kitchen (v1.3.1)
-----> Converging <default-ubuntu-1404>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 3.2.3...
       Removing non-cookbook files before transfer
```

Now when chef-zero starts up, it only has the nodes directory, but no clients or data bags:

``` bash
vagrant@default-ubuntu-1404:/tmp/kitchen$ pwd
/tmp/kitchen
vagrant@default-ubuntu-1404:/tmp/kitchen$ ls
backup  cache  client.pem  client.rb  cookbooks  dna.json  environments  nodes  validation.pem
vagrant@default-ubuntu-1404:/tmp/kitchen$ ls -l clients nodes
ls: cannot access clients: No such file or directory
nodes:
total 44
-rw-r--r-- 1 root root 42348 Feb  6 16:24 default-ubuntu-1404.json
```

There is already a client.pem, so the node doesn't re-register.  Now, when chef-vault does a node search and then tries to get the corresponding clients object for the node, it fails and the converge blows up:

```
       Recipe: chefci_tk_secrets::create_vaults
         * ruby_block[create vaults] action create
           ================================================================================
           Error executing action `create` on resource 'ruby_block[create vaults]'
           ================================================================================

           ChefVault::Exceptions::ClientNotFound
           -------------------------------------
           default-ubuntu-1404 is not a valid chef client and/or node
```
# Root Cause

The root cause is that TK is breaking the state set up by the first converge by deleting half of the bits that make up a node in chef.  A node is defined by the presence of two resources on the server: a node and a client.  Deleting the client renders the node unusable if you want to use the public key for anything.  I'll grant you that not many people do, but it is breaking the TK chef-repo.

I feel that you have to consider the nodes, clients and client.pem file to be atomic.  If you delete one of them you have to delete all of them because chef-client doesn't know how to (or in this case is incapable of) re-building from the broken half-state to a working system.
# Solution?

My first thought was just to remove clients from the list of dirs removed in the chef_base provisioner, but obviously this was added to an existing list of directories to clean for a reason.

I'm not sure what direction to go - either nodes and client.pem should also be removed (thus forcing the node to re-register on every converge, creating a new client object and keypair, or clients shouldn't be cleaned out.

Comments welcome

**Comment by jtimberman** [comment]
> changes were made to nuke and re-build the clients dir inside of /tmp/kitchen before converging.

That code is mostly copy/pasted from the `prepare_nodes` method. In fact if you look in that file in full, we have several "prepare" methods for the various directories that might contain fixture data for the "server." The code I wrote isn't doing anything special to remove the directory - as you mention, it's just part of test kitchen's "clean room" approach at the start of the run.

The specific reason I added the `clients` directory for test kitchen was to implement testing of the `chef-vault` cookbook itself :smile_cat:. In that cookbook, I use node/client fixtures for doing the testing. See here:

- https://github.com/opscode-cookbooks/chef-vault/tree/master/test/integration/clients
- https://github.com/opscode-cookbooks/chef-vault/blob/master/test/fixtures/cookbooks/test/recipes/chef_vault_secret.rb#L37-L42
- https://github.com/opscode-cookbooks/chef-vault/blob/master/test/integration/secret_resource/serverspec/secret_resource_spec.rb#L4

However, what if we remove the `/tmp/kitchen/client.pem` during the cleanup step?

**Comment by dougireton** [comment]
@jtimberman removing the /tmp/kitchen/client.pem during the cleanup step ought to do it. @jf647 what do you think?

**Comment by jf647** [comment]
Yes, removing the key should solve it.  TBH, moving away from my original approach for testing and using Josh's chef-vault cookbook helper has obviated the direct need, but it makes sense from a "let's not leave the chef-zero environment in a half-broken state" perspective.

**Comment by freimer** [comment]
Has there been any progress on this?  Indeed, doing a kitchen login and manually removing /tmp/kitchen/client.pem seems to "work," but it is not a sustainable solution.

**Comment by asachs01** [comment]
👍 for @freimer 's solution. Is there any further update on a long-term solution here?

**Comment by tas50** [comment]
We're not actively developing chef-vault at this point and we're suggesting that all users move away from the solution for secrets management with Chef Infra. Due to that we will not be improving how chef-vault is handled in Test Kitchen. If someone in the community wants to take this work on, we'd be more than happy to merge it. I'm going to close this issue out though since we won't be diving into the issues at this point.
